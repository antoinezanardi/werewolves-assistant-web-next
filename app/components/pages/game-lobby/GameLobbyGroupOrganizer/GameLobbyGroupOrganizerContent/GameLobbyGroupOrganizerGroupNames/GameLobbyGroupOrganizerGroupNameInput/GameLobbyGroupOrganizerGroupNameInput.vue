<template>
  <div id="game-lobby-group-organizer-group-name-input">
    <PrimeVueInplace>
      <template #display>
        <h5
          id="group-name"
          :aria-label="buttonAriaLabel"
        >
          <FontAwesomeIcon
            class="me-2 text-info"
            icon="edit"
          />

          <span>
            {{ trimmedLocalGroupName }}
          </span>
        </h5>
      </template>

      <template #content="{ closeCallback }">
        <PrimeVueInputGroup>
          <PrimeVueInputText
            id="group-name-input"
            v-model="localGroupName"
            :aria-label="$t('components.GameLobbyGroupOrganizerGroupNameInput.groupName')"
            autofocus
            :invalid="isButtonDisabled"
            :maxlength="30"
          />

          <PrimeVueButton
            id="submit-group-name-button"
            :aria-label="$t('components.GameLobbyGroupOrganizerGroupNameInput.submitGroupName')"
            :disabled="isButtonDisabled"
            severity="success"
            @click="setGroupName(closeCallback)"
          >
            <FontAwesomeIcon
              icon="check"
            />
          </PrimeVueButton>
        </PrimeVueInputGroup>
      </template>
    </PrimeVueInplace>
  </div>
</template>

<script setup lang="ts">
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import type { GameLobbyGroupOrganizerGroupNameInputEmits, GameLobbyGroupOrganizerGroupNameInputProps } from "~/components/pages/game-lobby/GameLobbyGroupOrganizer/GameLobbyGroupOrganizerContent/GameLobbyGroupOrganizerGroupNames/GameLobbyGroupOrganizerGroupNameInput/game-lobby-group-organizer-group-name-input.types";

const props = defineProps<GameLobbyGroupOrganizerGroupNameInputProps>();

const emit = defineEmits<GameLobbyGroupOrganizerGroupNameInputEmits>();

const { t } = useI18n();

const { groupName, otherGroupName } = toRefs(props);

const localGroupName = ref<string>(unref(groupName));

const trimmedLocalGroupName = computed<string>(() => localGroupName.value.trim());

const isButtonDisabled = computed<boolean>(() => !trimmedLocalGroupName.value || otherGroupName.value === trimmedLocalGroupName.value);

const buttonAriaLabel = computed<string>(() => t("components.GameLobbyGroupOrganizerGroupNameInput.editGroupName", { name: trimmedLocalGroupName.value }));

function setGroupName(closeCallback: () => void): void {
  localGroupName.value = trimmedLocalGroupName.value;
  emit("updateGroupName", trimmedLocalGroupName.value);
  closeCallback();
}
</script>