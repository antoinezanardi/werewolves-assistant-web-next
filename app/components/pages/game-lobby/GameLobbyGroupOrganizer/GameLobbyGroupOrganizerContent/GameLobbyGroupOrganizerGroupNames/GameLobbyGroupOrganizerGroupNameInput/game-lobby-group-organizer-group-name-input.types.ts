type GameLobbyGroupOrganizerGroupNameInputProps = {
  groupName: string;
  otherGroupName: string;
};

type GameLobbyGroupOrganizerGroupNameInputEmits = {
  updateGroupName: [string];
};

export type {
  GameLobbyGroupOrganizerGroupNameInputProps,
  GameLobbyGroupOrganizerGroupNameInputEmits,
};