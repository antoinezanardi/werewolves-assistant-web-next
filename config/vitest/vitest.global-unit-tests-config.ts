import { defineVitestConfig } from "@nuxt/test-utils/config";

import { VITEST_GLOBAL_UNIT_TESTS_CONFIG } from "./vitest.unit-tests-config.constants";

export default defineVitestConfig(VITEST_GLOBAL_UNIT_TESTS_CONFIG);