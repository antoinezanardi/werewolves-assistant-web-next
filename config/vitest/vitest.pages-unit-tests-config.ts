import { defineVitestConfig } from "@nuxt/test-utils/config";

import { VITEST_GLOBAL_UNIT_TESTS_CONFIG } from "./vitest.unit-tests-config.constants";

export default defineVitestConfig({
  ...VITEST_GLOBAL_UNIT_TESTS_CONFIG,
  test: {
    ...VITEST_GLOBAL_UNIT_TESTS_CONFIG?.test,
    include: [
      "./tests/unit/specs/pages/**/*.spec.ts",
      "./tests/unit/specs/app.nuxt.spec.ts",
      "./tests/unit/specs/error.nuxt.spec.ts",
    ],
  },
});