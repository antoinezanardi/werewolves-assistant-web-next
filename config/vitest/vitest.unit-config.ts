import { fileURLToPath } from "node:url";

import { defineVitestConfig } from "@nuxt/test-utils/config";

export default defineVitestConfig({
  test: {
    pool: "threads",
    root: fileURLToPath(new URL("../../", import.meta.url)),
    environment: "nuxt",
    environmentOptions: { nuxt: { rootDir: fileURLToPath(new URL("../../", import.meta.url)) } },
    setupFiles: ["./tests/unit/unit-setup.ts"],
    watch: false,
    include: ["./tests/unit/**/*.spec.ts"],
    coverage: {
      provider: "istanbul",
      exclude: [
        "nuxt.config.ts",
        "tests/**/*",
        "node_modules/**/*",
        "config/**/*",
        "plugins/**/*",
        "**/*.types.ts",
        "**/*.constants.ts",
      ],
      include: [
        "app/*.vue",
        "app/**/*.{vue,ts}",
      ],
      reportsDirectory: "./tests/unit/coverage",
      reporter: [
        "clover",
        "json",
        "lcov",
        "text",
        "text-summary",
        "html",
      ],
      all: true,
      thresholds: {
        lines: 100,
        functions: 100,
        branches: 100,
        statements: 100,
      },
    },
    globals: true,
    clearMocks: true,
    mockReset: true,
    restoreMocks: true,
  },
});