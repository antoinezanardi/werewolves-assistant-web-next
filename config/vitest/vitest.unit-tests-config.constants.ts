import { fileURLToPath } from "node:url";

import type { defineVitestConfig } from "@nuxt/test-utils/config";

const VITEST_GLOBAL_UNIT_TESTS_CONFIG: Parameters<typeof defineVitestConfig>[0] = {
  test: {
    pool: "threads",
    root: fileURLToPath(new URL("../../", import.meta.url)),
    environment: "nuxt",
    environmentOptions: {
      nuxt: {
        rootDir: fileURLToPath(new URL("../../", import.meta.url)),
        overrides: {
          runtimeConfig: {
            public: {
              defaultLocale: "en",
              werewolvesAssistantApi: { baseUrl: "http://127.0.0.1" },
              contactEmail: "john@doe.com",
            },
          },
        },
      },
    },
    setupFiles: ["./tests/unit/unit-setup.ts"],
    watch: false,
    include: ["./tests/unit/**/*.spec.ts"],
    coverage: {
      provider: "istanbul",
      exclude: [
        "nuxt.config.ts",
        "tests/**/*",
        "node_modules/**/*",
        "config/**/*",
        "app/plugins/**/*",
        "**/*.types.ts",
        "**/*.constants.ts",
      ],
      include: [
        "app/*.vue",
        "app/**/*.{vue,ts}",
      ],
      reportsDirectory: "./tests/unit/coverage",
      reporter: [
        "clover",
        "json",
        "lcov",
        "text",
        "text-summary",
        "html",
      ],
      all: true,
      thresholds: {
        lines: 100,
        functions: 100,
        branches: 100,
        statements: 100,
      },
    },
    globals: true,
    clearMocks: true,
    mockReset: true,
    restoreMocks: true,
  },
};

export { VITEST_GLOBAL_UNIT_TESTS_CONFIG };