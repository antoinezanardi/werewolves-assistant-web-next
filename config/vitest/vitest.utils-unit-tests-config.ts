import { defineVitestConfig } from "@nuxt/test-utils/config";

import { VITEST_GLOBAL_UNIT_TESTS_CONFIG } from "./vitest.unit-tests-config.constants";

export default defineVitestConfig({
  ...VITEST_GLOBAL_UNIT_TESTS_CONFIG,
  test: {
    ...VITEST_GLOBAL_UNIT_TESTS_CONFIG?.test,
    include: [
      "./tests/unit/specs/utils/*.spec.ts",
      "./tests/unit/specs/utils/**/*.spec.ts",
    ],
  },
});