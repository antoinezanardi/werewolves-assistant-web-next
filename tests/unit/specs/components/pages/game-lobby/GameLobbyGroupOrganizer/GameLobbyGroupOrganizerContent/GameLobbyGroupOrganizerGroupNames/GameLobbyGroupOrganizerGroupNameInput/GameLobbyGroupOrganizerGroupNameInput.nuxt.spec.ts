import type { mount } from "@vue/test-utils";

import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type Button from "primevue/button";
import type InputText from "primevue/inputtext";
import { beforeEach } from "vitest";
import type { GameLobbyGroupOrganizerGroupNameInputProps } from "~/components/pages/game-lobby/GameLobbyGroupOrganizer/GameLobbyGroupOrganizerContent/GameLobbyGroupOrganizerGroupNames/GameLobbyGroupOrganizerGroupNameInput/game-lobby-group-organizer-group-name-input.types";
import GameLobbyGroupOrganizerGroupNameInput from "~/components/pages/game-lobby/GameLobbyGroupOrganizer/GameLobbyGroupOrganizerContent/GameLobbyGroupOrganizerGroupNames/GameLobbyGroupOrganizerGroupNameInput/GameLobbyGroupOrganizerGroupNameInput.vue";

describe("Game Lobby Group Organizer Group Name Input Component", () => {
  const defaultProps: GameLobbyGroupOrganizerGroupNameInputProps = {
    groupName: "Group 1",
    otherGroupName: "Group 2",
  };
  let wrapper: ReturnType<typeof mount<typeof GameLobbyGroupOrganizerGroupNameInput>>;

  async function mountGameLobbyGroupOrganizerGroupNameInputComponent(options: ComponentMountingOptions<typeof GameLobbyGroupOrganizerGroupNameInput> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyGroupOrganizerGroupNameInput>>> {
    return mountSuspendedComponent(GameLobbyGroupOrganizerGroupNameInput, {
      props: defaultProps,
      global: {
        stubs: {
          Inplace: false,
          InputGroup: false,
        },
      },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyGroupOrganizerGroupNameInputComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should match snapshot when rendered without shallow rendering.", async() => {
    wrapper = await mountGameLobbyGroupOrganizerGroupNameInputComponent({ shallow: false });

    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Inplace content is not revealed", () => {
    it("should render the trimmed group name when rendered.", () => {
      const inplace = wrapper.find<HTMLDivElement>("#group-name");

      expect(inplace.text()).toBe(defaultProps.groupName);
    });
  });

  describe("Inplace content is revealed", () => {
    beforeEach(async() => {
      const button = wrapper.find<HTMLButtonElement>(".p-inplace-display");
      await button.trigger("click");
    });

    it("should match snapshot without shallow rendering when inplace content is revealed.", async() => {
      wrapper = await mountGameLobbyGroupOrganizerGroupNameInputComponent({ shallow: false });
      const button = wrapper.find<HTMLButtonElement>(".p-inplace-display");
      await button.trigger("click");

      expect(wrapper).toBeTruthy();
      expect(wrapper.html()).toMatchSnapshot();
    });

    describe("Group name input", () => {
      it("should be invalid when group name is empty.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue("");
        await input.trigger("blur");

        expect(input.attributes("invalid")).toBe("true");
      });

      it("should be invalid when group name is the same as the other group name.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue(defaultProps.otherGroupName);
        await input.trigger("blur");

        expect(input.attributes("invalid")).toBe("true");
      });

      it("should be invalid when group name is the same as the other group name even with start and end whitespace.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue(`  ${defaultProps.otherGroupName} `);
        await input.trigger("blur");

        expect(input.attributes("invalid")).toBe("true");
      });

      it("should have a max length of 30 characters when rendered.", () => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");

        expect(input.attributes("maxlength")).toBe("30");
      });
    });

    describe("Group name submit button", () => {
      it("should set button to disabled when group name is empty.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue("");
        await input.trigger("blur");
        const button = wrapper.findComponent<typeof Button>("#submit-group-name-button");

        expect(button.attributes("disabled")).toBe("true");
      });

      it("should set button to disabled when group name is the same as the other group name.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue(defaultProps.otherGroupName);
        await input.trigger("blur");
        const button = wrapper.findComponent<typeof Button>("#submit-group-name-button");

        expect(button.attributes("disabled")).toBe("true");
      });

      it("should set button to disabled when group name is the same as the other group name even with start and end whitespace.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue(`  ${defaultProps.otherGroupName} `);
        await input.trigger("blur");
        const button = wrapper.findComponent<typeof Button>("#submit-group-name-button");

        expect(button.attributes("disabled")).toBe("true");
      });

      it("should emit updateGroupName event when button is clicked.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue("New Group 1");
        const button = wrapper.findComponent<typeof Button>("#submit-group-name-button");
        await button.trigger("click");

        expect(wrapper.emitted("updateGroupName")).toBeTruthy();
        expect(wrapper.emitted("updateGroupName")?.[0][0]).toBe("New Group 1");
      });

      it("should set input value to new group name when button is clicked.", async() => {
        const input = wrapper.findComponent<typeof InputText>("#group-name-input");
        await input.setValue("New Group 1");
        const button = wrapper.findComponent<typeof Button>("#submit-group-name-button");
        await button.trigger("click");

        expect(input.attributes("modelvalue")).toBe("New Group 1");
      });
    });
  });
});