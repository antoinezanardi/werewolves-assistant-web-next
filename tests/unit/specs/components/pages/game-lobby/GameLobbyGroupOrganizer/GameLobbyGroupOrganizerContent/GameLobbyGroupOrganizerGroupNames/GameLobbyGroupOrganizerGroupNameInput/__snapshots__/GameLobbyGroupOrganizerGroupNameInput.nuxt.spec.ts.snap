// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Group Organizer Group Name Input Component > Inplace content is revealed > should match snapshot without shallow rendering when inplace content is revealed. 1`] = `
"<div id="game-lobby-group-organizer-group-name-input">
  <div class="p-inplace p-component" aria-live="polite" data-pc-name="inplace" data-pc-section="root" pc7="">
    <div class="p-inplace-content" data-pc-section="content">
      <div class="p-inputgroup" data-pc-name="inputgroup" data-pc-section="root" pc10=""><input type="text" class="p-inputtext p-component p-filled" id="group-name-input" autofocus="" maxlength="30" data-pc-name="inputtext" data-pc-section="root" value="Group 1" pc8=""><button class="p-button p-component p-button-success" type="button" data-pc-name="button" data-p-disabled="false" data-p-severity="success" id="submit-group-name-button" data-pc-section="root" data-pd-ripple="true" style="overflow: hidden; position: relative;" pc9=""><svg class="svg-inline--fa fa-check" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="check" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
            <path class="" fill="currentColor" d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"></path>
          </svg><span role="presentation" aria-hidden="true" data-p-ink="true" data-p-ink-active="false" class="p-ink" pd1="" data-pc-name="ripple" data-pc-section="root"></span></button></div>
    </div>
  </div>
</div>"
`;

exports[`Game Lobby Group Organizer Group Name Input Component > should match snapshot when rendered without shallow rendering. 1`] = `
"<div id="game-lobby-group-organizer-group-name-input">
  <div class="p-inplace p-component" aria-live="polite" data-pc-name="inplace" data-pc-section="root" pc3="">
    <div class="p-inplace-display" tabindex="0" role="button" data-pc-section="display">
      <h5 id="group-name"><svg class="svg-inline--fa fa-edit me-2 text-info" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="edit" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
          <path class="" fill="currentColor" d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160L0 416c0 53 43 96 96 96l256 0c53 0 96-43 96-96l0-96c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 96c0 17.7-14.3 32-32 32L96 448c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l96 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 64z"></path>
        </svg><span>Group 1</span></h5>
    </div>
  </div>
</div>"
`;

exports[`Game Lobby Group Organizer Group Name Input Component > should match snapshot when rendered. 1`] = `
"<div id="game-lobby-group-organizer-group-name-input">
  <div class="p-inplace p-component" aria-live="polite" data-pc-name="inplace" data-pc-section="root" pc1="">
    <div class="p-inplace-display" tabindex="0" role="button" data-pc-section="display">
      <h5 id="group-name">
        <font-awesome-icon-stub icon="edit" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" class="me-2 text-info"></font-awesome-icon-stub><span>Group 1</span>
      </h5>
    </div>
  </div>
</div>"
`;
